### PR TITLE
docs: clarify the assumption on numeric values of JSON parsers we make

### PR DESCRIPTION
**Pull request by poettering**
Prompted by:

https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/message/BOBD6KVTXPR6K5ANAX6LIJLKNSGXCR3B/

### CODE REVIEW

**Comment by bluca** [comment]
Not too familiar with this - just to check, does this requirement pose any issue to the common Linux parsers (json-c, jansson, glib-json) or are they all already compliant?

File: docs/COREDUMP_PACKAGE_METADATA.md [hunk 1]
```diff
@@ -67,6 +67,14 @@ Value: a JSON string with the structure described below
 }
 ```
 
+When it comes to JSON numbers, this specification assumes that JSON parsers
+processing this information are capable of reproducing the full signed and
+unsigned 64bit integer range (i.e. -2⁶³…+2⁶⁴-1) as well as the full 64bit IEEE
```
Comment (fweimer-rh) [inline docs/COREDUMP_PACKAGE_METADATA.md]: The full 64-bit range is definitely not supported everywhere:

```javascript
const json = '{"number": 9007199254740993}';
const obj = JSON.parse(json);
console.log(obj.number);
```
This comes back as: `9007199254740992`.

So you should restrict this to 52-bit integers with a separate sign.

And of course many implementations are not able to round-trip floating pointer numbers accurately through decoding and encoding (either changing the textual representation to something ugly, or not encoding the value accurately).

Comment (bluca) [reply docs/COREDUMP_PACKAGE_METADATA.md]: I doubt we'll ever need all of that for this purpose, so IMHO we should specify the lowest common denominator among the current and popular oss implementations

Comment (poettering) [reply docs/COREDUMP_PACKAGE_METADATA.md]: I htink we have the liberty here to make choice. If this was somethin we expect to be consumed by JS then yeah, I think we should make this more restrictive. But given this is mostly expected to be processed by system-level stuff, and 64bit integers are kinda a given for that these days I'd not limit things needlessly here. but dunno, i am fine with anything as long as we document it.

if we didn't document it then the only thing people reasonably can expect is the 52bit integer thing. But if we declare it I think we can be more free.

Comment (poettering) [reply docs/COREDUMP_PACKAGE_METADATA.md]: (the json libs i checked are all system level stuff, C, python, glib, and so on)

**Comment by poettering** [comment]
> Not too familiar with this - just to check, does this requirement pose any issue to the common Linux parsers (json-c, jansson, glib-json) or are they all already compliant?

glib-json only does int64_t, doesn't cover uint64_t

**Comment by bluca** [comment]
> > Not too familiar with this - just to check, does this requirement pose any issue to the common Linux parsers (json-c, jansson, glib-json) or are they all already compliant?
> 
> glib-json and jansson only does int64_t, doesn't cover uint64_t. json-c and our own does full uint64_t range. python does arbitrary range integers

I'd say restrict the wording to int64_t then? It would seem to me in this case we have no need at the moment for any of these extremes, but we are trying to get folks convinced to adopt the specification, so there's an advantage in being able to include as many tools as possible

**Comment by keszybz** [comment]
> I'd say restrict the wording to int64_t then?

+1. All we really care about is to avoid trouble some numbers being misinterpreted. And since we don't need any large numbers, just limiting the allowed range to something that is easily supported by all known implementations solves the problem.

File: docs/COREDUMP_PACKAGE_METADATA.md (hunk 1 above)
Comment (fweimer-rh) [reply docs/COREDUMP_PACKAGE_METADATA.md]: The node.js tooling may need to generate/process this data, too. Javascript isn't just for browsers anymore.

Just restrict it to 52 bits, please.

**Comment by keszybz** [comment]
It seems all reviewers agree — please change it to `int64_t` so we can call it a day.

**Comment by fweimer-rh** [comment]
> It seems all reviewers agree — please change it to `int64_t` so we can call it a day.

Maybe I do not count as a reviewer, but assuming 64-bit integer support in JSON implementations is clearly incorrect. [RFC 8259](https://datatracker.ietf.org/doc/html/rfc8259#page-8) also cautions about this:

> Note that when such software is used, numbers that are integers and
> are in the range [-(2\**53)+1, (2**53)-1] are interoperable in the
> sense that implementations will agree exactly on their numeric
> values.

**Comment by keszybz** [comment]
Oh, sorry. My goal was to also take your review into account, i.e. to say to define this as the lowest common denominator, but somehow my brain rounded this up to the nearest commonly-used value.

**Comment by poettering** [comment]
So I still think we should not restrict ourselves needlessly, given we define the spec. i.e. I am not convinced that compat with JS is really what we should focus on here, and as long as we declare it we should be fine making stricter, yet reasonable requirements.

But apparently, I am the only one thinking this. I thus used the documented range from RFC8259. ptal.

**Review by bluca — APPROVED** [review]
lgtm

**Pull request merged**